Update `visit_source_value` `JOIN`s to only use `hadm_id`

briangow opened the pull request:
This PR should be merged after the PRs ahead of it. 

In #57 , we updated `visit_source_value` to simply be `hadm_id` instead of a concatenation of `hadm_id` and other variables. This PR updates the `JOIN` statements which use `visit_source_value` to simply `JOIN` against `hadm_id`, see 225c79c7e2252377bc595ec65a0d2d994e87bb96.

briangow committed:
update joins on visit_source_value

atsvetkova-ody commented on etl/etl/cdm_drug_exposure.sql:
```diff
@@ -57,34 +57,34 @@ CREATE OR REPLACE TABLE @etl_project.@etl_dataset.cdm_drug_exposure
-    src.source_concept_id                       AS drug_source_concept_id,
-    src.route_source_code                       AS route_source_value,
-    src.dose_unit_source_code                   AS dose_unit_source_value,
+    `@etl_project.@etl_dataset`.obf_id_str(src.trace_id, 32)    AS drug_exposure_id,
```
I'm a little bit worried about uniqueness of drug_exposure_id when the generated value is based on src.trace_id only: drug_exposure source rows can be multiplied due to multiple mapping of drug_source_value (for example, when a drug mapped to ingredients concepts). Maybe src.trace_id + drug_concept_id would be a safer choice?

atsvetkova-ody reviewed (commented):
I looked through the changes, and I have only two notes:
1) about drug_exposure_id: I expect some risk of generating duplicated values when only src.trace_id is passed to the obfuscating function. May be src.trace_id + drug_concept_id could be a safer option.
2) about visit_source_value, just an idea probably to think it over later: we used to populate visit_source_value with the source visit_id to track original visits. But having trace_id field or with adding another non-OMOP field dedicated to source visit_id, it can be safe to return initial role to visit_source_value, i. e. to populate it with the values used to find visit_concept_id.

briangow committed:
ensure drug_exposure_id is unique

briangow commented:
Thanks for the feedback @atsvetkova-ody ! 

> about drug_exposure_id: I expect some risk of generating duplicated values when only src.trace_id is passed to the obfuscating function. May be src.trace_id + drug_concept_id could be a safer option.

I've updated this to combine `trace_id` with `drug_concept_id`/`src.target_concept_id` to ensure this is unique. 

> about visit_source_value, just an idea probably to think it over later: we used to populate visit_source_value with the source visit_id to track original visits. But having trace_id field or with adding another non-OMOP field dedicated to source visit_id, it can be safe to return initial role to visit_source_value, i. e. to populate it with the values used to find visit_concept_id.

I don't think I understand what you are suggesting here. Looking at the old code, on main, `visit_source_value` is being set based on a number of different variables depending on the source table ( as seen in `AS source_value` in https://github.com/OHDSI/MIMIC/blob/main/etl/etl/lk_vis_part_2.sql ). The thought for updating this was that since `hadm_id` is unique per visit, that it cleanly represents the `visit_source_value`. 

Can I ask if you've reviewed all of the open PR commits (most of which are in this PR since I've been building on top of PRs to avoid merge conflicts)? I ask so I can decide what to merge.

atsvetkova-ody commented:
> I don't think I understand what you are suggesting here. Looking at the old code, on main, `visit_source_value` is being set based on a number of different variables depending on the source table ( as seen in `AS source_value` in https://github.com/OHDSI/MIMIC/blob/main/etl/etl/lk_vis_part_2.sql ). The thought for updating this was that since `hadm_id` is unique per visit, that it cleanly represents the `visit_source_value`.
> 
> Can I ask if you've reviewed all of the open PR commits (most of which are in this PR since I've been building on top of PRs to avoid merge conflicts)? I ask so I can decide what to merge.

Hi Brian, I'm sorry for the pause, these two weeks were a little crazy!

I confirm that I reviewed all PR commits two weeks ago, and the new one about drug_exposure_id now. Everything looks good for me.

About my suggestion on `visit_source_value`: my suggestion was to populate visit_source_value with string values like "Inpatient", "Telehealth" etc., moving hadm_id to a new non-omop custom field in visit_occurrence -- this is our current best practice. This suggestion is not about fixing an error, it is more like improvement from good to better.

briangow committed:
replace visit_source_value joins

briangow commented:
As discussed above and updated in #57 , we changed how `visit_source_value` is set. It use to be set based on `subject_id` and `hadm_id`. Later in the build we would `JOIN` `visit_source_value` (from visit tables) with `subject_id` and `hadm_id` (in `src` tables). Since this will no longer work, I've updated these `JOIN`s to:

1. `ON  vis.subject_id = src.subject_id AND vis.hadm_id = src.hadm_id` in cases where `hadm_id` is always present
2. `ON  vis.subject_id = src.subject_id AND (vis.hadm_id = src.hadm_id OR vis.hadm_id IS NULL AND vis.date_id = src.date_id)` in cases where `hadm_id` might not be present. 

@atsvetkova-ody , please let me know if there is anything I missed.

atsvetkova-ody commented:
Hi @briangow , I'm confirming that everything is good.
